File: server/services/goodsService.js
/**
 * 商品服务层：封装商品查询逻辑
 * 控制器只做参数校验与响应，业务查询集中在这里
 */
const { Op } = require('sequelize')
const { Goods, Category } = require('../models')
const AppError = require('../utils/appError')

/**
 * 分页查询商品列表
 * @param {object} params { categoryId, keyword, page, pageSize }
 * @returns {object} { list, total, page, pageSize, hasMore }
 */
async function listGoods({ categoryId, keyword, page = 1, pageSize = 10 }) {
  const where = { status: true } // 只展示上架商品

  if (categoryId) {
    where.category_id = categoryId
  }
  if (keyword) {
    where.name = { [Op.like]: `%${keyword}%` } // 名称模糊搜索
  }

  const { count, rows } = await Goods.findAndCountAll({
    where,
    order: [['sales', 'DESC']], // 默认按销量排序
    offset: (page - 1) * pageSize,
    limit: pageSize,
    attributes: { exclude: ['images', 'detail'] }, // 列表不返回大字段，详情接口单独取
  })

  return {
    list: rows,
    total: count,
    page,
    pageSize,
    hasMore: page * pageSize < count,
  }
}

/**
 * 商品详情（含轮播图与图文详情）
 */
async function getGoodsDetail(id) {
  const goods = await Goods.findByPk(id, {
    include: [{ model: Category, as: 'category', attributes: ['id', 'name'] }],
  })
  if (!goods) {
    throw new AppError('商品不存在或已下架', 404, 404)
  }
  if (!goods.status) {
    throw new AppError('商品已下架', 404, 404)
  }

  // 解析 JSON 字段（数据库存的是字符串）
  const data = goods.toJSON()
  try {
    data.images = JSON.parse(data.images || '[]')
  } catch {
    data.images = []
  }
  return data
}

/**
 * 热销商品（推荐位/首页横向滚动）
 */
async function listHot(limit = 10) {
  return Goods.findAll({
    where: { status: true, is_hot: true },
    order: [['sales', 'DESC']],
    limit,
    attributes: { exclude: ['images', 'detail'] },
  })
}

/**
 * 新品商品
 */
async function listNew(limit = 10) {
  return Goods.findAll({
    where: { status: true, is_new: true },
    order: [['created_at', 'DESC']],
    limit,
    attributes: { exclude: ['images', 'detail'] },
  })
}

module.exports = { listGoods, getGoodsDetail, listHot, listNew }

File: springboot/src/main/java/com/freshmall/service/GoodsService.java
package com.freshmall.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.freshmall.common.AppException;
import com.freshmall.entity.Category;
import com.freshmall.entity.Goods;
import com.freshmall.repository.CategoryRepository;
import com.freshmall.repository.GoodsRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 商品服务（对应 Node 版 services/goodsService.js）
 */
@Service
@RequiredArgsConstructor
public class GoodsService {

    private final GoodsRepository goodsRepository;
    private final CategoryRepository categoryRepository;
    private final ObjectMapper objectMapper; // Spring 注入（已注册 JSR310 时间模块）

    /**
     * 分页商品列表（列表不返回大字段 images/detail）
     * 返回 { list, total, page, pageSize, hasMore }
     */
    public Map<String, Object> listGoods(Integer categoryId, String keyword, int page, int pageSize) {
        String like = (keyword == null || keyword.isBlank()) ? null : keyword.trim();
        Pageable pageable = PageRequest.of(page - 1, pageSize);
        var result = goodsRepository.search(categoryId == 0 ? null : categoryId, like, pageable);

        List<Map<String, Object>> list = result.getContent().stream().map(this::toListMap).toList();
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("list", list);
        data.put("total", result.getTotalElements());
        data.put("page", page);
        data.put("pageSize", pageSize);
        data.put("hasMore", page * pageSize < result.getTotalElements());
        return data;
    }

    /**
     * 商品详情（含轮播图数组与分类名）
     */
    public Map<String, Object> getGoodsDetail(Integer id) {
        Goods goods = goodsRepository.findById(id)
                .orElseThrow(() -> new AppException("商品不存在或已下架", 404, 404));
        if (!goods.getStatus()) {
            throw new AppException("商品已下架", 404, 404);
        }

        Map<String, Object> data = toDetailMap(goods);
        // 解析轮播图 JSON
        try {
            List<String> images = objectMapper.readValue(
                    goods.getImages() == null ? "[]" : goods.getImages(),
                    new TypeReference<List<String>>() {});
            data.put("images", images);
        } catch (Exception e) {
            data.put("images", new ArrayList<>());
        }
        // 关联分类
        Category category = categoryRepository.findById(goods.getCategory_id()).orElse(null);
        if (category != null) {
            Map<String, Object> catMap = new LinkedHashMap<>();
            catMap.put("id", category.getId());
            catMap.put("name", category.getName());
            data.put("category", catMap);
        }
        return data;
    }

    public List<Map<String, Object>> listHot(int limit) {
        return goodsRepository.findHotGoods(PageRequest.of(0, limit))
                .stream().map(this::toListMap).toList();
    }

    public List<Map<String, Object>> listNew(int limit) {
        return goodsRepository.findNewGoods(PageRequest.of(0, limit))
                .stream().map(this::toListMap).toList();
    }

    /** 列表视图（不含大字段） */
    private Map<String, Object> toListMap(Goods g) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", g.getId());
        m.put("category_id", g.getCategory_id());
        m.put("name", g.getName());
        m.put("image", g.getImage());
        m.put("price", g.getPrice());
        m.put("original_price", g.getOriginal_price());
        m.put("unit", g.getUnit());
        m.put("stock", g.getStock());
        m.put("sales", g.getSales());
        m.put("is_hot", g.getIs_hot());
        m.put("is_new", g.getIs_new());
        m.put("status", g.getStatus());
        m.put("createdAt", g.getCreatedAt());
        m.put("updatedAt", g.getUpdatedAt());
        return m;
    }

    /** 详情视图（全字段） */
    private Map<String, Object> toDetailMap(Goods g) {
        Map<String, Object> m = toListMap(g);
        m.put("images", g.getImages());
        m.put("detail", g.getDetail());
        return m;
    }
}
